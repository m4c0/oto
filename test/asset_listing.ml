let asset_of_script ({ actor; pose; _ } : Oto.Types.opcode) =
  let ch = match pose with None -> actor | Some p -> actor ^ "-" ^ p in
  "char/" ^ ch

let rec assets_of (acc : string list) = function
  | [] -> List.sort_uniq String.compare acc
  | sc :: scl -> assets_of_meta acc scl sc

and assets_of_meta (acc : string list) (scl : Oto.Types.scene list)
    (sc : Oto.Types.scene) =
  let ({ background; music; script; _ } : Oto.Types.scene_meta), next = sc () in
  let bg = Option.map (Filename.concat "back") background in
  let music = Option.map (Filename.concat "music") music in
  let ch = List.map asset_of_script script in
  let res =
    List.map Option.to_list [ bg; music ]
    |> List.flatten |> List.append ch |> List.append acc
  in
  match next with
  | Continuation s | WaitThenJump s -> assets_of_meta res scl s
  | Choice l -> List.map Oto.Types.scene_of_choice l |> assets_of res
  | EndGame -> assets_of res scl

let () = assets_of [] [ Poc.game ] |> List.iter print_endline
