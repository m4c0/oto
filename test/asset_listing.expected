back/end
back/restaurant
back/title
char/lefty
char/midly
char/midly-hot
char/righty
music/game-over
music/romance-1
